### PR TITLE
feat: SLO resource

### PR DESCRIPTION
**Pull request by roshan8**

### CODE REVIEW

File: internal/api/client.go [hunk 1]
```diff
@@ -126,6 +126,14 @@ func RequestSlice[TReq any, TRes any](method string, url string, client *Client,
 	return *data, nil
 }
 
+func Get[TRes interface{}](client *Client, ctx context.Context, path string) (*TRes, error) {
+	return Request[interface{}, TRes](http.MethodGet, path, client, ctx, nil)
+}
+
```
Comment (dheeraj-squadcast) [inline internal/api/client.go]: Please revert, we have other ways of calling get and post methods

File: internal/provider/resource_slo.go [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+				Optional:    true,
+				Default:     "team",
+			},
+			"team_id": {
+				Description:  "The team which SLO resource belongs to",
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: tf.ValidateObjectID,
+			},
+			"org_id": {
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go]: Not required, see client.OrgID

File: internal/provider/resource_slo.go [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+						"team_id": {
+							Description: "Team ID.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+					},
+				},
+				Optional: true,
+			},
+			"owner_type": {
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go]: remove this, it's always going to be "team", let's not burden the user

File: internal/provider/resource_slo.go [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+							Computed:    true,
+						},
+						"team_id": {
+							Description: "Team ID.",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+					},
+				},
+				Optional: true,
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go]: move this up, difficult to see

File: internal/provider/resource_slo.go [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+							},
+							Optional: true,
+						},
+						"service": {
+							Description:  "The ID of the service in which the user want to create an incident",
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: tf.ValidateObjectID,
+						},
+						"owner_type": {
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go]: remove

File: internal/provider/resource_slo.go [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Description: "The ID of the notification rule",
+							Type:        schema.TypeInt,
+							Computed:    true,
+						},
+						"slo_id": {
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go:142]: no need to have this, this is got from slo resource itself

File: internal/provider/resource_slo.go [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+								"Only supported for rules name \"increased_false_positives\" and \"remaining_error_budget\"",
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"is_checked": {
+							Description: "Is checked?",
+							Type:        schema.TypeBool,
+							Computed:    true,
+						},
+						"owner_type": {
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go]: remove

File: internal/provider/resource_slo.go [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+				Description: "SLO monitoring checks has rules for monitoring any SLO violation(Or warning signs)",
+				Type:        schema.TypeList,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Description: "The ID of the monitoring rule",
+							Type:        schema.TypeInt,
+							Computed:    true,
+						},
+						"slo_id": {
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go:100]: remove

File: internal/provider/resource_slo.go [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+			"duration_in_days": {
+				Description: "Tracks SLO for the last x days. Required only when SLO time interval type set to \"rolling\"",
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Computed:    true,
+			},
+			"start_time": {
+				Description:  "SLO start time. Required only when SLO time interval type set to \"fixed\"",
+				Type:         schema.TypeString,
+				Computed:     true,
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go:79]: why optional and computed?

File: internal/provider/resource_slo.go [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+				Optional:     true,
+				ValidateFunc: validation.IsRFC3339Time,
+			},
+			"end_time": {
+				Description:  "SLO end time. Required only when SLO time interval type set to \"fixed\"",
+				Type:         schema.TypeString,
+				Computed:     true,
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go:86]: why optional and computed?

File: internal/provider/resource_slo.go [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+			},
+			"time_interval_type": {
+				Description:  "Type of the SLO. Values can either be \"rolling\" or \"fixed\"",
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringInSlice([]string{"rolling", "fixed"}, false),
+			},
+			"duration_in_days": {
+				Description: "Tracks SLO for the last x days. Required only when SLO time interval type set to \"rolling\"",
+				Type:        schema.TypeInt,
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go:72]: why optional and computed?

File: internal/provider/resource_slo.go [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+							Description: "The ID of the notification rule",
+							Type:        schema.TypeInt,
+							Computed:    true,
+						},
+						"slo_id": {
+							Description: "The ID of the SLO.",
+							Type:        schema.TypeInt,
+							Computed:    true,
+						},
+						"users": {
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go]: inconsistent naming - if it's id, suffix with _id, and if plural, _ids

File: internal/provider/resource_slo.go [hunk 13]
```diff
@@ -0,0 +1,426 @@
...
+							Description: "List of user ID's who should be alerted via email.",
+							Type:        schema.TypeList,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+							Optional: true,
+						},
+						"squads": {
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go]: inconsistent naming - if it's id, suffix with _id, and if plural, _ids

File: internal/provider/resource_slo.go [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+							Description: "List of Squad ID's who should be alerted via email.",
+							Type:        schema.TypeList,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+							Optional: true,
+						},
+						"service": {
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go]: inconsistent naming - if it's id, suffix with _id, and if plural, _ids

File: internal/provider/resource_slo.go [hunk 15]
```diff
@@ -0,0 +1,426 @@
...
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: tf.ValidateObjectID,
+						},
+						"owner_type": {
+							Description: "The Owner type",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"team_id": {
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go]: remove,

File: internal/provider/resource_slo.go [hunk 16]
```diff
@@ -0,0 +1,426 @@
...
+							Description: "Is checked?",
+							Type:        schema.TypeBool,
+							Computed:    true,
+						},
+						"owner_type": {
+							Description: "SLO owner type",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+						"team_id": {
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go]: remove

File: internal/provider/resource_slo.go [hunk 17]
```diff
@@ -0,0 +1,426 @@
...
+				Description: "Owner type",
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "team",
+			},
+			"team_id": {
+				Description:  "The team which SLO resource belongs to",
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: tf.ValidateObjectID,
```
Comment (dheeraj-squadcast) [inline internal/provider/resource_slo.go:176]: set forcenew = true

File: internal/tf/state.go [hunk 18]
```diff
@@ -9,11 +9,16 @@ import (
 func SetState(d *schema.ResourceData, m map[string]any) error {
 	id, ok := m["id"].(string)
 	if !ok {
-		return fmt.Errorf("invalid id")
+		// if resource ID is an integer
+		idInt, ok := m["id"].(uint)
+		if !ok {
```
Comment (dheeraj-squadcast) [inline internal/tf/state.go]: there are nested ifs + duplicated setids

replace the whole thing with a type switch followed by a SetID

File: docs/data-sources/squadcast_squad.md [hunk 19]
```diff
@@ -0,0 +1,28 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "squadcast_squad Data Source - terraform-provider-squadcast-new"
+subcategory: ""
```
Comment (dheeraj-squadcast) [inline docs/data-sources/squadcast_squad.md]: please remove docs from this PR

**Review by dheeraj-squadcast — CHANGES_REQUESTED** [review]

File: internal/provider/resource_slo.go (hunk 9 above)
Comment (roshan8) [reply internal/provider/resource_slo.go:79]: If SLO is of type fixed then the user will enter `start date` and `end date`
When SLO is of type rolling then this field is not considered.

File: internal/provider/resource_slo.go (hunk 12 above)
Comment (roshan8) [reply internal/provider/resource_slo.go]: sure, corrected

File: internal/provider/resource_slo.go (hunk 13 above)
Comment (roshan8) [reply internal/provider/resource_slo.go]: done

File: internal/provider/resource_slo.go (hunk 14 above)
Comment (roshan8) [reply internal/provider/resource_slo.go]: done

File: internal/api/client.go (hunk 1 above)
Comment (roshan8) [reply internal/api/client.go]: done

File: internal/provider/resource_slo.go (hunk 2 above)
Comment (roshan8) [reply internal/provider/resource_slo.go]: done

File: internal/provider/resource_slo.go (hunk 3 above)
Comment (roshan8) [reply internal/provider/resource_slo.go]: removed

File: internal/provider/resource_slo.go (hunk 4 above)
Comment (roshan8) [reply internal/provider/resource_slo.go]: done

File: internal/provider/resource_slo.go (hunk 5 above)
Comment (roshan8) [reply internal/provider/resource_slo.go]: done

File: internal/provider/resource_slo.go (hunk 6 above)
Comment (roshan8) [reply internal/provider/resource_slo.go:142]: done

File: internal/provider/resource_slo.go (hunk 7 above)
Comment (roshan8) [reply internal/provider/resource_slo.go]: done

File: internal/provider/resource_slo.go (hunk 8 above)
Comment (roshan8) [reply internal/provider/resource_slo.go:100]: done

Comment (roshan8) [reply internal/provider/resource_slo.go:100]: The user not required to add this now, being added before making the call.

File: internal/provider/resource_slo.go (hunk 10 above)
Comment (roshan8) [reply internal/provider/resource_slo.go:86]: If SLO is of type fixed then the user will enter start date and end date
When SLO is of type rolling then this field is not considered.

File: internal/provider/resource_slo.go (hunk 11 above)
Comment (roshan8) [reply internal/provider/resource_slo.go:72]: its required if slo type is rolling, we need to pass duration_in_days.
and API response contains `start, end time and duration in days` no matter whether slo is fixed or rolling.

File: internal/provider/resource_slo.go (hunk 15 above)
Comment (roshan8) [reply internal/provider/resource_slo.go]: done

File: internal/provider/resource_slo.go (hunk 16 above)
Comment (roshan8) [reply internal/provider/resource_slo.go]: done, owner_id only required at the top level now

File: internal/provider/resource_slo.go (hunk 17 above)
Comment (roshan8) [reply internal/provider/resource_slo.go:176]: done

File: internal/tf/state.go (hunk 18 above)
Comment (roshan8) [reply internal/tf/state.go]: changed

File: docs/data-sources/squadcast_squad.md (hunk 19 above)
Comment (roshan8) [reply docs/data-sources/squadcast_squad.md]: removed content from docs directory

**Pull request merged**